atom/browser/api/save_page_handler.cc

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2015 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "atom/browser/api/save_page_handler.h"
+
+#include <string>
+
+#include "atom/browser/atom_browser_context.h"
+#include "atom/browser/atom_download_manager_delegate.h"
+#include "base/callback.h"
+#include "base/files/file_path.h"
+#include "content/public/browser/web_contents.h"
+
+namespace atom {
+
+namespace api {
+
+SavePageHandler::SavePageHandler(content::WebContents* web_contents,
+                                 const SavePageCallback& callback)
+    : id_(0),
+      web_contents_(web_contents),
+      callback_(callback) {
+}
+
+SavePageHandler::~SavePageHandler() {
+}
+
+void SavePageHandler::OnDownloadCreated(content::DownloadManager* manager,
+                                        content::DownloadItem* item) {
+  if (id_ == item->GetId()) {
+    GetDownloadManager()->RemoveObserver(this);
+    item->AddObserver(this);
+  }
+}
+
+bool SavePageHandler::Handle(const base::FilePath& full_path,
+                             const content::SavePageType& save_type) {
+  auto delegate = static_cast<AtomDownloadManagerDelegate*>(
+      GetDownloadManager()->GetDelegate());
+  GetDownloadManager()->AddObserver(this);
```

Comment:
The `AddObserver` code is needed here, it gives a chance for `SavePageHandler` to listen `OnDownloadCreated` event (The `DownloadItem` is created during the `SavePage` calling).